tensorflow/compiler/xla/service/gpu/target_util.cc

```diff
@@ -18,85 +18,289 @@ limitations under the License.
+      if (target_triple.getArch() == llvm::Triple::nvptx ||
+          target_triple.getArch() == llvm::Triple::nvptx64) {
+        return TargetInfo(nvptx_intrinsic_info, default_nvptx_function_info,true);
+      } else {
+        return TargetInfo(amdgpu_intrinsic_info,default_amdgpu_function_info, true);
+      }
     }
-    case TargetIntrinsicID::kBlockIdz: {
-      return {llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_z,
-              llvm::Intrinsic::amdgcn_workgroup_id_z};
+    case TargetFunctionID::kBlockIdz: {
+      TargetIntrinsicInfo nvptx_intrinsic_info(llvm::Intrinsic::nvvm_read_ptx_sreg_ctaid_z);
+      TargetIntrinsicInfo amdgpu_intrinsic_info(llvm::Intrinsic::amdgcn_workgroup_id_z);
+      if (target_triple.getArch() == llvm::Triple::nvptx ||
+          target_triple.getArch() == llvm::Triple::nvptx64) {
+        return TargetInfo(nvptx_intrinsic_info, default_nvptx_function_info,true);
+      } else {
+        return TargetInfo(amdgpu_intrinsic_info,default_amdgpu_function_info, true);
+      }
     }
-    case TargetIntrinsicID::kBarrierId: {
-      return {llvm::Intrinsic::nvvm_barrier0,
-              llvm::Intrinsic::amdgcn_s_barrier};
+    case TargetFunctionID::kBarrierId: {
+      TargetIntrinsicInfo nvptx_intrinsic_info(llvm::Intrinsic::nvvm_barrier0);
+      TargetIntrinsicInfo amdgpu_intrinsic_info(llvm::Intrinsic::amdgcn_s_barrier);
+      if (target_triple.getArch() == llvm::Triple::nvptx ||
+          target_triple.getArch() == llvm::Triple::nvptx64) {
+        return TargetInfo(nvptx_intrinsic_info, default_nvptx_function_info,true);
+      } else {
+        return TargetInfo(amdgpu_intrinsic_info,default_amdgpu_function_info, true);
+      }
     }
   }
 }
 }  // namespace
 
-llvm::CallInst* EmitCallToTargetIntrinsic(
-    TargetIntrinsicID intrinsic_id, absl::Span<llvm::Value* const> operands,
-    absl::Span<llvm::Type* const> overloaded_types, llvm::IRBuilder<>* b) {
+llvm::Value* EmitCallToTargetFunction(
+    TargetFunctionID function_id, absl::Span<llvm::Value* const> operands,
+    absl::Span<const PrimitiveType> input_types, PrimitiveType output_type,
+    absl::Span<const llvm::Attribute::AttrKind> attributes,
+    absl::Span<llvm::Type* const> overloaded_types,
+    llvm::IRBuilder<>* b) {
   llvm::Module* module = b->GetInsertBlock()->getModule();
-  struct TargetIntrinsics gpu_intrinsic_id = GetIntrinsic(intrinsic_id);
   llvm::Triple target_triple = llvm::Triple(module->getTargetTriple());
-  llvm::Intrinsic::ID llvm_intrinsic_id = llvm::Intrinsic::not_intrinsic;
+  struct TargetInfo gpu_info = GetTargetInfo(function_id, target_triple);
 
-  if ((target_triple.getArch() == llvm::Triple::nvptx) ||
-      (target_triple.getArch() == llvm::Triple::nvptx64)) {
-    llvm_intrinsic_id = gpu_intrinsic_id.nvptx_intrinsic;
-  } else if (target_triple.getArch() == llvm::Triple::amdgcn) {
-    llvm_intrinsic_id = gpu_intrinsic_id.amdgpu_intrinsic;
+  if (gpu_info.has_intrinsic) {
```

Comment:
@sumesh13 shall `gpu_info.target_intrinsic_info` and `gpu_info.target_function_info` becomes `absl::optional`, we don't really need `gpu_info.has_intrinsic` field?